### ISSUE TITLE
Play data seems incorrect

### ISSUE THREAD

**Issue by brbeaird** [derek-adair/nflgame#34]
Looking over the data from 2018 week 1, I'm seeing some things that don't quite line up. Check out play 3673 in 2018091001.json. This play was a 2 yard gain for Todd Gurley. However, he is listed further down in one of the sequence nodes with 32 yards.

While this may not be visible in nflgame, that data flows into nfldb as incorrect data, so if you're trying to do any aggregate queries there, they will not be correct.

Any ideas here?

![image](https://user-images.githubusercontent.com/7239272/45372793-74e8fe00-b5b3-11e8-8c6d-12743fbad914.png)

**Comment by 19eddiedean19** [comment]
I have a few other examples of this I noticed as well. When i get back to a computer i will try and get examples from the json.

There were a lot of fumbles being counted twice, including: Christian McCaffrey, Alex Collins, Ben Roethlisberger, Dalvin Cook, Alfred Morris, Dak Prescott, Trubisky. 
Looking in nfldb - play_player table their respective drives actually have 2 fumbles counted.
I have not yet had a chance to look into it

 player_id  |  gsis_id   | play_id | fumbles_lost | team 
------------+------------+---------+--------------+------
 00-0033869 | 2018090912 |    3871 |            2 | CHI
 00-0033077 | 2018090910 |    3931 |            2 | DAL
 00-0033280 | 2018090910 |     443 |            2 | CAR
 00-0022924 | 2018090901 |    3867 |            2 | PIT
 00-0030433 | 2018090906 |    1730 |            2 | NO
 00-0033893 | 2018090904 |    1163 |            2 | MIN
 00-0029141 | 2018090904 |    1590 |            2 | SF

**Comment by brbeaird** [comment]
What's the best way to delete those games and regenerate? And do you know if there is a separate step for cleaning up nfldb?

**Comment by ochawkeye** [comment]
In Windows I just went `explorer \my\path\to\python\lib\site-packages\nflgame\gamecenter-json`, sorted by name, and deleted everything from `2018090600.json` through `2018091001.json`

Yes, I see that `nfldb` retains the "bad" Sandejo information despite `nflgame` being now accurate

```
# nflgame:
(SF, MIN 1, Q2, 2 and 1) (2:30) A.Morris up the middle to MIN 1 for no gain (L.Joseph). FUMBLES (L.Joseph), RECOVERED by MIN-H.Smith at MIN 2. H.Smith to MIN 2 for no gain (P.Garcon).
H.Smith OrderedDict([('defense_frec_yds', 0), ('defense_frec', 1)])
P.Garcon OrderedDict([('defense_tkl', 1)])
A.Morris OrderedDict([('rushing_att', 1), ('rushing_yds', -1), ('fumbles_tot', 1), ('fumbles_forced', 1), ('fumbles_lost', 1)])
L.Joseph OrderedDict([('defense_tkl', 1), ('defense_ffum', 1)])
```
```
# nfldb:
(SF, OPP 1, Q2, 2 and 1) (2:30) A.Morris up the middle to MIN 1 for no gain (L.Joseph). FUMBLES (L.Joseph), RECOVERED by MIN-H.Smith at MIN 2. H.Smith to MIN 2 for no gain (P.Garcon).
Linval Joseph (MIN, DT) {'defense_tkl': 1, 'defense_ffum': 1}
Alfred Morris (SF, RB) {'fumbles_tot': 2, 'fumbles_forced': 2, 'rushing_att': 1, 'rushing_yds': -1, 'fumbles_lost': 2}
Pierre Garcon (SF, WR) {'defense_tkl': 2}
Andrew Sendejo (MIN, DB) {'defense_tkl': 2, 'defense_ffum': 2}
Harrison Smith (MIN, DB) {'defense_frec_yds': 1, 'defense_frec': 2}
```

I don't really know the *best* way to delete the games from `nfldb`.  You can always access the database itself, but not super convenient, especially if you've not done it before.

One thing you can do is run this little script, and then run `nfldb-update`.
```python
import nfldb
db = nfldb.connect()

gsis_ids = [2018090600, 2018090900, 2018090901, 2018090902, 2018090903, 
            2018090904, 2018090905, 2018090906, 2018090907, 2018090908, 
            2018090909, 2018090910, 2018090911, 2018090912, 2018091000, 
            2018091001]

for gsis_id in gsis_ids:
    query = "DELETE FROM game where gsis_id = '{}';".format(gsis_id)
    with nfldb.Tx(db) as cursor:
        cursor.execute(query)  
```
That will fix `nfldb`
```
(SF, OPP 1, Q2, 2 and 1) (2:30) A.Morris up the middle to MIN 1 for no gain (L.Joseph). FUMBLES (L.Joseph), RECOVERED by MIN-H.Smith at MIN 2. H.Smith to MIN 2 for no gain (P.Garcon).
Linval Joseph (MIN, DT) {'defense_tkl': 1, 'defense_ffum': 1}
Alfred Morris (SF, RB) {'fumbles_tot': 1, 'fumbles_forced': 1, 'rushing_att': 1, 'rushing_yds': -1, 'fumbles_lost': 1}
Pierre Garcon (SF, WR) {'defense_tkl': 1}
Harrison Smith (MIN, DB) {'defense_frec': 1}
```

**Comment by mickeelm** [comment]
If this is a very rare thing, then maybe it's not worth it but these resources do send the Last-Modified HTTP header so it's possible to have a mechanism checking for changes after a game is finished by sending HEAD requests. Where to store that info (the timestamp) I'm not sure but that would probably be the cheapest way.

**Comment by derek-adair** [comment]
Seems like its worth addressing to me.

Reading this `Last-Modified` seems elegant enough to me!  Why not just store that directly on the game json?

It seems pretty low impact to me...

**Comment by ochawkeye** [comment]
I noticed this happened again in last night's BAL/CIN game.  CIN was being
credited with an extra INT, 2 extra fumble recoveries, and 1.5 extra sacks
banged on the play-by-play double when the last play of the game occurred.
At some point between last night and this morning that data was fixed and
deleting my local .json and regenerating fixed the discrepancy.

**Comment by brbeaird** [comment]
If this is going to happen every week, I'd say it's a fairly significant problem.

**Comment by brbeaird** [comment]
Yeah that makes sense.

**Comment by mickeelm** [comment]
Storing the timestamp in the game-json would indeed be low impact. What needs to be handled in that case is support for reading both (older) files without the timestamp and files with the timestamp. Easy to fix.

What needs to be decided is when/how long to check for updates of games. For instance, we don't need to check the correctness of a game played in 2016. Games of current week, x weeks back...? 

So the proposed logic would be to check `Last-Modified` with HEAD requests, and if a resource has changed, perform a GET request and replace the game file.

**Comment by derek-adair** [comment]
Hmmmm very good point!  I need to look at how this is done to provide a more informed opinion.... but perhaps an easy/non-obtrusive way would be to only check if the game occured within 2 weeks.

IMO the way to ensure the most accuracy would be to check EVERY game.  But i'm unsure if this would result in too much traffic that the NFL may crack down.

**Comment by derek-adair** [comment]
Actually we could get away with simply changing the logic and instead of checking / updating currently playing games... we update all games in the past two weeks.  I'm not so sure tracking last-modified header does much if we need to keep track of how recntly the game was played anyways.

**Comment by ochawkeye** [comment]
If I'm understanding correctly, on Monday nights, everytime someone accesses the API we will be dumping redownloading ~31 games worth of data instead of the one game being played? I hit the API roughly every 3 minutes (and I think nfldb is every 30 seconds).  That seems like a lot of unnecessary calls for data that is almost certainly static.  I think that might be a bit too lazy of a solution here.

**Comment by mickeelm** [comment]
In practice, after analyzing how much traffic that goes just by opening NFL.com, seeing how many page visits they have every month, I don't think that we really need to worry about calling them too often (again, in practice). With that said, I think it's **great** that we are cautious and try to be smart. Not only does that decrease the number of calls to the NFL, it makes for a better application that runs smoother.

> I'm not so sure tracking last-modified header does much if we need to keep track of how recntly the game was played anyways.

We have the schedule on file/in memory. Getting entries from a certain relative time range is an easy operation so no worries there.

Come to think of it, we could make use of an extra flag like `verified` or something like that. That is, if a game is checked at earliest x days (5 days..?) after it was played, it's pretty safe to say that we won't need to check that anymore. Hence, we can flag it `verified`. Maybe check (with HEAD) about once a day up "until x".

A user that runs the application pretty much all the time would have calls scattered over time, whereas a user that hasn't run the application in say three weeks will have to download everything anyway and all games older than (today - x) will automatically be `verified`. 

For the historical data, it's not impossible that we have errors so one could of course verify these as well if we wanted to (preferably a single user that commits all changes to the repo).

**Comment by derek-adair** [comment]
I wonder if we should make this thing configurable, ya know?

VERIFY_POST_GAME=(False|interval)

or perhaps a manual call.  Like... some kind of force update function.

**Comment by derek-adair** [comment]
Again, i'm not 100% sure on how the player stats are updated, but ASSUMING it's done in a batch process of games that are active and removed from the dictionary of games to update once they are complete.

_VERIFY_POST_GAME_ could be checked whenever we look up if a game is active or not.  If the game was marked active w/in the interval we continue checking this.

I believe this would prevent those extraneous calls @ochawkeye mentioned.

**Comment by ochawkeye** [comment]
Yes, this is a data problem that occurred again this week.  Just looking at defense's accumulated stats, I saw changes in the following after deleting and regenerating stats:

Baltimore: 4 sacks > 3 sacks
Buffalo: 4 fumble recoveries > 2
Carolina: 2 interceptions > 1
Chicago: 3 fumble recoveries > 2
Cincinnati: 2 sacks > 1
Cleveland: 3 interceptions > 2, 3 sacks > 2
Houston: 5 sacks > 4
Jacksonville: 1 fumble recovery > 0 (this one seems weird), 4 sacks > 3
Kansas City: 5 sacks > 4
Minnestoa: 4 sacks > 3
New York Jets: 5 sacks > 4
Seattle: 3 interceptions > 2

This is not a problem that I ever witnessed before this season.

**Comment by ochawkeye** [comment]
In regards to being a high priority bug, I'm not so certain on that one.  Just because the underlying JSON data shows the redundant play information, that doesn't mean that their game center is displaying all of that information.  I haven't watched the game's live in the game center, but it's probable that their displaying of the information is handling the redundancies properly.

**Comment by brbeaird** [comment]
Yep, this is the first season I've seen this, and I looked at things pretty closely last year.

It does seem like a bug, but seeing as how the NFL feed eventually gets corrected, that may just be how they've got it working now. I have no idea what's going on behind the scenes there, but it definitely feels like something we'll end up having to deal with to make sure we go back and get the corrected data at some point.

**Comment by derek-adair** [comment]
Regardless of any redundancies there will be corrections after the game is no longer active.  The doubly counted plays may or may not be corrected post-game-closing.  I'm all about the VERIFY_POST_GAME config option.

Also note that this project is basically running itself so "high priority" is a relative term.

**Comment by ochawkeye** [comment]
"High Priority" was in the context of whether or not NFL.com viewed this as a bug in their source data, not a grade of how quickly `nflgame` should be addressing it.

**Comment by ochawkeye** [comment]
Storing the last pulled information in the JSON would be fine, but even that might not be necessary.  All comes down to how long after the game the data ends up being corrected.

I've found if a change is going to happen it happens by the morning after.  But I have not compared stats from any other site to see if they match what we are recording so I can't say with 100% certainty that after the morning after no changes occur anymore.    If we could collectively agree that it looks like the morning after is the cutoff for changes, then we could probably just tap into the metadata for the .json.gz file itself.  If the file was last modified on the day of the game that happened in the past, then delete it and pull it again.  If it is newer than the game date then call it good.

As mentioned, these post-game changes are not something we have seen previously.  Each week the NFL has a handful of plays they reclassify based on further analysis.  For example, there might be a pass play that, after further tape review, gets classified as a running play instead.  In that case, they publish a list of statistical changes for the play (ie. Drew Brees: -1 completion, -1 pass attempt, -6 passing yards; Mark Ingram: -1 reception, -6 receiving yards, +1 rushing attempt, +6 rushing yards).  But those stat changes have never been retroactively applied to the play-by-play data.  So these post-game changes are all pretty new behavior for those of us that have been using `nflgame` for years.

**Comment by brbeaird** [comment]
I know with most Fantasy sites, stat corrections usually go through on Wed
or Thurs morning for the previous week's games; I'm not sure the longest
possible time NFL could make changes after the fact, though.

**Comment by mickeelm** [comment]
So after some thinking...

Storing the `Last-Modified`-value in the `json`-files isn't really necessary since we can use the timestamps of the files themselves for the same purpose, they would give the same info pretty much in the application's point of view. 

Also, each time we check a remote game file on nfl.com (via `HEAD`) we can update the corresponding local file _even ifthe game has not changed_, e.g. even if there's no download, but just to indicate that this game has been checked at this very timestamp. So after checking games postgame for x number of days x times per day, this would kind of work like a "verified-flag". E.g. if the local modified-timestamp is past x days from game day or whatever (or x days/hours from the assumed correction point, like wed or thu morning) then that would be considered a verified game, we don't need to check anymore.

No tampering with the json-files, they're all just copies of the nfl.com data. I kinda like that approach.

What we haven't touched is _where_ this should happen, e.g. _where in the code_. We've discussed this in prior issues. "Hi-jacking" `import nflgame` like with the schedule?

**Comment by ochawkeye** [comment]
>  What we haven't touched is where this should happen, e.g. where in the code. We've discussed this in prior issues. "Hi-jacking" import nflgame like with the schedule?

Main reason I don't like that is that it is possible to run nflgame without ever accessing any of the games.  The following is perfectly valid code.

```python
import nflgame
print len(nflgame.players)
```
```
8631
```
  Adding the overhead of checking every single game (thousands) against when it appears in the schedule seems overly aggressive.

Instead of determining if we need to _replace_ the .JSON, what if we just delay caching the .JSON until some point after the game has concluded? 

https://github.com/derek-adair/nflgame/blob/master/nflgame/game.py#L307-L309

It would mean pulling the same data from nfl.com over and over again until we hit that magic date/time when we say we're satisfied that changes won't be happening anymore.

Just throwing it out there, not suggesting it is the best solution to the problem.

**Comment by mickeelm** [comment]
> Main reason I don't like that is that it is possible to run nflgame without ever accessing any of the games. The following is perfectly valid code.

I don't like it either, in fact I think it's poor design. I don't think that `import nflgame` should do anything but import the lib. I think that all external as well as internal calls should be initiated by the user. Using the `live`-module would be such a thing, the user makes an active choice.

I've just gotten the impression that we are OK with (like?) that for instance the schedule is automatically updated with the import so I've just kinda followed that philosophy. Personally I don't like it but I can live with it if it's what the team wishes :slightly_smiling_face: 

> Adding the overhead of checking every single game (thousands) against when it appears in the schedule seems overly aggressive.

Unclear of me, I was thinking maybe the last x week(s) but didn't write it down. A user initiated function call/script could override this to make checks earlier in time.

> Instead of determining if we need to replace the .JSON, what if we just delay caching the .JSON until some point after the game has concluded?

Best solution so far I think. This is easy to fix for now and solves the problem. Well done! The only thing I'm not sure about is if _all_ games get synced at the same time. Like TNF vs MNF...so would it be a set time (like, thursdays at 8AM) or a relative time (like 24 hours after kickoff).

**Comment by derek-adair** [comment]
Marking on hold b/c #46 will address this if we are lucky

**Comment by derek-adair** [comment]
Also yes, updating should not be done on import and only when you run the live script should it update.

I've come to see the error in my thinking and understanding of the code base.

**Comment by derek-adair** [comment]
@brbeaird I've updated [nfldb](https://github.com/derek-adair/nfldb/) to use python3 and have a[ compatible database](https://github.com/derek-adair/nfldb/releases/download/1.0.0a4/nfldb2019.sql.zip) w/ (hopefully) all of the 2019 data.

**Comment by derek-adair** [comment]
[/feeds-rs/playbyplay/](http://www.nfl.com/feeds-rs/playbyplay/1998090611) has data back to 1998.  We can use this to verify the accuracy of our play data.  I will be writing a script to do this.

I'm not sure the best way to invoke this in real time to ensure accuracy.  I'm considering adding config for stuff like this where you set your preference as a user:

config.ini example:
```
[nflgame]
# Period in which to check the game json
verify_play_data = never[default] | hourly | daily | weekly | monthly
# update the schedule on import
update_schedule_on_import= false[default] | true
```

**Comment by brbeaird** [comment]
> @brbeaird I've updated [nfldb](https://github.com/derek-adair/nfldb/) to use python3 and have a[ compatible database](https://github.com/derek-adair/nfldb/releases/download/1.0.0a4/nfldb2019.sql.zip) w/ (hopefully) all of the 2019 data.
> 
> 
> 
> 

Ooh nice. I’ll check that out when I get a chance. Eventually want to get that in a docker.

**Comment by derek-adair** [comment]
[it ALREADY ISSSSSSSSSSSSSS!](https://github.com/derek-adair/nfldb/blob/master/docker-compose.yml)
